packages/cmus/aaudio.patch

```diff
@@ -0,0 +1,994 @@
++static int op_aaudio_unpause(void)
++{
++	// request stream start, wait until it starts to start (i.e., will start
++	// consuming frames written to it)
++	return -OP_ERROR_AAUDIO(do_state_change(AAudioStream_requestStart, AAUDIO_STREAM_STATE_STARTED, AAUDIO_STREAM_STATE_STARTING));
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_buffer_space(void)
++{
++	// optimal buffer amount (anecdotally, this generally seems to be less
++	// than half the max buffer amount)
++	return AAudioStream_getFramesPerBurst(strm) * strm_frame_size;
++
++	// max buffer amount (without blocking)
++	// return AAudioStream_getBufferSizeInFrames(strm) * strm_frame_size;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_init(void)
++{
++	if (!aaudio_supported()) {
++		// skip the output plugin (see op_select_any)
++		return -OP_ERROR_NOT_SUPPORTED;
++	}
++	return OP_ERROR_SUCCESS;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_exit(void)
++{
++	return OP_ERROR_SUCCESS;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_open(int *volume_max)
++{
++	*volume_max = UINT16_MAX;
++
++	return OP_ERROR_SUCCESS;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_close(void)
++{
++	return OP_ERROR_SUCCESS;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_get_fds(int what, int *fds)
++{
++	switch (what) {
++	case MIXER_FDS_OUTPUT:
++		fds[0] = mixer_notify_output_out;
++		return 1;
++	default:
++		return 0;
++	}
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_set_volume(int l, int r)
++{
++	return -OP_ERROR_NOT_SUPPORTED;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++static int op_aaudio_mixer_get_volume(int *l, int *r)
++{
++	// aaudio doesn't support volume control, so say the volume is 100%
++	*l = *r = UINT16_MAX;
++
++	return OP_ERROR_SUCCESS;
++}
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++const struct output_plugin_ops op_pcm_ops = {
++	.init = op_aaudio_init,
++	.exit = op_aaudio_exit,
++	.open = op_aaudio_open,
++	.close = op_aaudio_close,
++	.drop = op_aaudio_drop,
++	.write = op_aaudio_write,
++	.pause = op_aaudio_pause,
++	.unpause = op_aaudio_unpause,
++	.buffer_space = op_aaudio_buffer_space,
++};
++
++REQUIRES_API(AAUDIO_MINIMUM_API)
++const struct mixer_plugin_ops op_mixer_ops = {
++	.init = op_aaudio_mixer_init,
++	.exit = op_aaudio_mixer_exit,
++	.open = op_aaudio_mixer_open,
++	.close = op_aaudio_mixer_close,
++	.get_fds.abi_2 = op_aaudio_mixer_get_fds,
++	.set_volume = op_aaudio_mixer_set_volume,
++	.get_volume = op_aaudio_mixer_get_volume,
++};
++
++const struct output_plugin_opt op_pcm_options[] = {
++	OPT(op_aaudio, performance_mode),
++	OPT(op_aaudio, allowed_capture),
++	OPT(op_aaudio, sharing_mode),
++	OPT(op_aaudio, disable_spatialization),
++	{ NULL },
++};
++
++const struct mixer_plugin_opt op_mixer_options[] = {
++	{ NULL },
++};
++
++const int op_priority = -3; // higher priority than pulse (-2)
```

Comment:
I set this to be a higher priority than pulse since it should almost always be the better option if it's supported.

Another option could be to split the pulseaudio plugin into a subpackage, with the aaudio plugin having a lower priority.